### PR TITLE
[ST Enhancement] Deletion of remaining pvcs after tests.

### PR DESCRIPTION
**Pull request by jankalinic**
Signed-off-by: jkalinic <jkalinic@redhat.com>

### Type of change

_Select the type of your PR_

- Bugfix
- Enhancement / new feature
- Refactoring
- Documentation

### Description

This PR consists of simple enhancement for system tests. PVCs that have annotation 'deletionClaim' set to false, won't be deleted by default after the tests. So this PR contains deletion strategy for those PVCs as we want to clear every resources afterwards.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: test/src/main/java/io/strimzi/test/k8s/KubeClient.java [hunk 1]
```diff
@@ -232,6 +232,14 @@ public List<String> listPodNamesInSpecificNamespace(String namespaceName, String
             .collect(Collectors.toList());
     }
 
+    public PersistentVolumeClaim getPVC(String namespaceName, String pvcName) {
```
Comment (see-quick) [inline test/src/main/java/io/strimzi/test/k8s/KubeClient.java:235]: Camel case and same for delete
```suggestion
    public PersistentVolumeClaim getPvc(String namespaceName, String pvcName) {
```

File: systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java [hunk 2]
```diff
@@ -66,12 +67,24 @@ public static void waitUntilPVCAnnotationChange(String namespaceName, String clu
...
 
-    public static void waitForPVCDeletion(String namespaceName, int volumesCount, JbodStorage jbodStorage, String clusterName) {
+    public static void waitForPVCDeletion(String namespaceName, String pvcName) {
+        TestUtils.waitFor("Wait for PVC deletion", Constants.POLL_INTERVAL_FOR_RESOURCE_DELETION, Duration.ofMinutes(3).toMillis(), () -> {
+            if (kubeClient(namespaceName).getPVC(namespaceName, pvcName) == null) {
+                return true;
+            } else {
+                LOGGER.warn("PVC {} is not deleted yet! Triggering force delete by cmd client!", pvcName);
+                cmdKubeClient(namespaceName).deleteByName("pvc", pvcName);
+                return false;
```
Comment (see-quick) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java:77]: Do we need to trigger this operation multiple times? :) Are there some scenarios, where PVCs are stuck or why do need it? :)

File: test/src/main/java/io/strimzi/test/k8s/KubeClient.java [hunk 3]
```diff
@@ -232,6 +232,14 @@ public List<String> listPodNamesInSpecificNamespace(String namespaceName, String
...
+        return client.persistentVolumeClaims().inNamespace(namespaceName).withName(pvcName).get();
+    }
+
+    public boolean deletePVC(String namespaceName, String pvcName) {
```
Comment (see-quick) [inline test/src/main/java/io/strimzi/test/k8s/KubeClient.java:239]:
```suggestion
    public boolean deletePvc(String namespaceName, String pvcName) {
```

**Review by see-quick — COMMENTED** [review]
Overall looks good, I just have a few nits.

File: test/src/main/java/io/strimzi/test/k8s/KubeClient.java (hunk 1 above)
Comment (im-konge) [reply test/src/main/java/io/strimzi/test/k8s/KubeClient.java:235]: Why camel case here? It's an abbreviation for `PersistentVolumeClaim`. For `StrimziPodSet` it's also SPS.

File: systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java (hunk 2 above)
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java:77]: do we have this in "force" delete command `--force --grace-period=0 ` ? It might speed up the deletion a lot

File: systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java [hunk 4]
```diff
@@ -66,12 +67,24 @@ public static void waitUntilPVCAnnotationChange(String namespaceName, String clu
...
+        });
+    }
+
+    public static void waitForJbodDeletion(String namespaceName, int volumesCount, JbodStorage jbodStorage, String clusterName) {
         int numberOfPVCWhichShouldBeDeleted = jbodStorage.getVolumes().stream().filter(
             singleVolumeStorage -> ((PersistentClaimStorage) singleVolumeStorage).isDeleteClaim()
         ).collect(Collectors.toList()).size();
 
-        TestUtils.waitFor("Wait for PVC deletion", Constants.POLL_INTERVAL_FOR_RESOURCE_DELETION, Duration.ofMinutes(6).toMillis(), () -> {
+        TestUtils.waitFor("Wait for JBOD deletion", Constants.POLL_INTERVAL_FOR_RESOURCE_DELETION, Duration.ofMinutes(6).toMillis(), () -> {
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java:87]:
```suggestion
        TestUtils.waitFor("Wait for JBOD storage deletion", Constants.POLL_INTERVAL_FOR_RESOURCE_DELETION, Duration.ofMinutes(6).toMillis(), () -> {
```

mozna zmenit ja nazev te metody?

File: systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java (hunk 2 above)
Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java:77]: Anyway, it should be part of different PR, because it could influence much more parts than current changes

File: systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java [hunk 5]
```diff
@@ -66,12 +67,24 @@ public static void waitUntilPVCAnnotationChange(String namespaceName, String clu
...
 
-    public static void waitForPVCDeletion(String namespaceName, int volumesCount, JbodStorage jbodStorage, String clusterName) {
+    public static void waitForPvcDeletion(String namespaceName, String pvcName) {
+        TestUtils.waitFor("Wait for PVC deletion", Constants.POLL_INTERVAL_FOR_RESOURCE_DELETION, Duration.ofMinutes(3).toMillis(), () -> {
+            if (kubeClient(namespaceName).getPvc(namespaceName, pvcName) == null) {
+                return true;
+            } else {
+                LOGGER.warn("PVC {} is not deleted yet! Triggering force delete by cmd client!", pvcName);
+                cmdKubeClient(namespaceName).deleteByName("pvc", pvcName);
+                return false;
```
Comment (im-konge) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java:77]:
```suggestion
            if (kubeClient(namespaceName).getPvc(namespaceName, pvcName) != null)  {
                LOGGER.warn("PVC {} is not deleted yet! Triggering force delete by cmd client!", pvcName);
                cmdKubeClient(namespaceName).deleteByName("pvc", pvcName);
                return false;
            }
            
            return true;
```

File: test/src/main/java/io/strimzi/test/k8s/KubeClient.java [hunk 6]
```diff
@@ -232,6 +232,14 @@ public List<String> listPodNamesInSpecificNamespace(String namespaceName, String
...
+        return client.persistentVolumeClaims().inNamespace(namespaceName).withName(pvcName).get();
+    }
+
+    public boolean deletePvc(String namespaceName, String pvcName) {
+        return client.persistentVolumeClaims().inNamespace(namespaceName).withName(pvcName).delete();
+    }
+
     public List<PersistentVolumeClaim> listPersistentVolumeClaims(String namespaceName, String clusterName) {
         return client.persistentVolumeClaims().inNamespace(namespaceName).list().getItems().stream()
             .filter(persistentVolumeClaim -> persistentVolumeClaim.getMetadata().getName().contains(clusterName))
```
Comment (im-konge) [inline test/src/main/java/io/strimzi/test/k8s/KubeClient.java:245]: One method has name `listPersistentVolumeClaims`, other two have `getPvc` or `deletePvc` -> I would either keep the abbreviation or long name, not both.

**Review by see-quick — APPROVED** [review]

File: test/src/main/java/io/strimzi/test/k8s/KubeClient.java [hunk 7]
```diff
@@ -232,6 +232,14 @@ public List<String> listPodNamesInSpecificNamespace(String namespaceName, String
             .collect(Collectors.toList());
     }
 
+    public PersistentVolumeClaim getPersistentVolumeClaim(String namespaceName, String pvcName) {
+        return client.persistentVolumeClaims().inNamespace(namespaceName).withName(pvcName).get();
+    }
+
+    public boolean deletePersistentVolumeClaims(String namespaceName, String pvcName) {
```
Comment (im-konge) [inline test/src/main/java/io/strimzi/test/k8s/KubeClient.java:239]:
```suggestion
    public boolean deletePersistentVolumeClaim(String namespaceName, String pvcName) {
```

File: systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java [hunk 8]
```diff
@@ -66,12 +67,24 @@ public static void waitUntilPVCAnnotationChange(String namespaceName, String clu
         LOGGER.info("PVC annotation has changed {}", newAnnotation.toString());
     }
 
-    public static void waitForPVCDeletion(String namespaceName, int volumesCount, JbodStorage jbodStorage, String clusterName) {
+    public static void waitForPvcDeletion(String namespaceName, String pvcName) {
+        TestUtils.waitFor("Wait for PVC deletion", Constants.POLL_INTERVAL_FOR_RESOURCE_DELETION, Duration.ofMinutes(3).toMillis(), () -> {
+            if (kubeClient(namespaceName).getPersistentVolumeClaim(namespaceName, pvcName) != null) {
```
Comment (im-konge) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java:72]:
```suggestion
            if (kubeClient().getPersistentVolumeClaim(namespaceName, pvcName) != null) {
```
you can remove it from the `kubeClient()` itself, as it will take only the `namespaceName` from the method parameter (in `getPersistentVolumeClaim`)

File: systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java [hunk 9]
```diff
@@ -66,12 +67,24 @@ public static void waitUntilPVCAnnotationChange(String namespaceName, String clu
         LOGGER.info("PVC annotation has changed {}", newAnnotation.toString());
     }
 
-    public static void waitForPVCDeletion(String namespaceName, int volumesCount, JbodStorage jbodStorage, String clusterName) {
+    public static void waitForPvcDeletion(String namespaceName, String pvcName) {
+        TestUtils.waitFor("Wait for PVC deletion", Constants.POLL_INTERVAL_FOR_RESOURCE_DELETION, Duration.ofMinutes(3).toMillis(), () -> {
```
Comment (im-konge) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/kubeUtils/objects/PersistentVolumeClaimUtils.java:71]: Also, can we take a timeout from `Constants` or, if there will be any matching, add it there this one?

**Review by im-konge — APPROVED** [review]
Few small nits. LGTM otherwise :)

File: systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java [hunk 10]
```diff
@@ -74,6 +77,12 @@ public void delete(Kafka resource) {
 
         kafkaClient().inNamespace(namespaceName).withName(
             resource.getMetadata().getName()).withPropagationPolicy(DeletionPropagation.FOREGROUND).delete();
+
+        // additional deletion of pvcs with specification deleteClam set to false which were not deleted prior this method
```
Comment (michalxo) [inline systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java:81]: deleteClam

File: systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java [hunk 11]
```diff
@@ -74,6 +77,12 @@ public void delete(Kafka resource) {
 
         kafkaClient().inNamespace(namespaceName).withName(
             resource.getMetadata().getName()).withPropagationPolicy(DeletionPropagation.FOREGROUND).delete();
+
+        // additional deletion of pvcs with specification deleteClaim set to false which were not deleted prior this method
+        for (PersistentVolumeClaim pvc : kubeClient().listPersistentVolumeClaims(namespaceName, clusterName)) {
+            kubeClient().deletePersistentVolumeClaim(namespaceName, pvc.getMetadata().getName());
+            PersistentVolumeClaimUtils.waitForPersistentVolumeClaimDeletion(namespaceName, pvc.getMetadata().getName());
+        }
```
Comment (scholzj) [inline systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java:85]: Why not do it in two cycles?
1) Request deletion of all PVCs in one loop
2) Wait for deletion in the second loop

That way, if PVC deletion takes 10 seconds and you delete 10 PVCs, you will nto wait 100 seconds but only ~10 seconds.

**Review by scholzj — COMMENTED** [review]
I left on nit. But I wonder why this is needed. AFAIU the tests now run in a namespace per test-suite. And the namespace should be delete after the test suite is complete. That should delete the PVCs as well. So why is this needed in addition to that?

File: systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java (hunk 11 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java:85]: > I left on nit. But I wonder why this is needed. AFAIU the tests now run in a namespace per test-suite. And the namespace should be delete after the test suite is complete. That should delete the PVCs as well. So why is this needed in addition to that?

Because, there are edge cases, where for instance you install Strimzi by OLM. When you install Strimzi by OLM and you want to have a cluster-wide setting, it has to be installed in the specific namespace where global `OperatorGroup` is installed. Such namespace we do not delete because it would also delete global `OperatorGroup` and thus we wouldn't be able to deploy Operator.

**Comment by scholzj** [comment]
> Because, there are edge cases, where for instance you install Strimzi by OLM. When you install Strimzi by OLM and you want to have a cluster-wide setting, it has to be installed in the specific namespace where global OperatorGroup is installed. Such namespace we do not delete because it would also delete global OperatorGroup and thus we wouldn't be able to deploy Operator.

@see-quick But I would not expect you to deploy any Kafka clusters into the OLM namespace with cluster-wide operators, or? So there should not be any PVCs in there.

**Comment by see-quick** [comment]
> @see-quick But I would not expect you to deploy any Kafka clusters into the OLM namespace with cluster-wide operators, or? So there should not be any PVCs in there.

We could possibly change that but currently, we are deploying resources inside that namespace when we execute almost all @IsolatedSuites (which does not create its own namespace).

File: systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java (hunk 11 above)
Comment (jankalinic) [reply systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java:85]: re two cycles: I haven't bumped into a situation where deletion takes this much time. If there is a problem with deletion it's always caught in waitingForDeletion. Deletion is almost instantly completed in every situation. So in my opinion it should not be necessary.

Comment (scholzj) [reply systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java:85]: If it is instant, you do not need to wait for it ... if you need to wait for it, you should split it to save time ;-)

**Review by Frawless — APPROVED** [review]

File: systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java (hunk 11 above)
Comment (jankalinic) [reply systemtest/src/main/java/io/strimzi/systemtest/resources/crd/KafkaResource.java]: You are right, I'm adding proposed solution for time save. Thank you for the suggestion.

**Review by scholzj — APPROVED** [review]
